processes/record_clusterer.py

```diff
@@ -0,0 +1,202 @@
+    MAX_MATCH_DISTANCE = 4
+    CLUSTER_SIZE_LIMIT = 10000
+    IDENTIFIERS_TO_MATCH = r'\|(?:isbn|issn|oclc|lccn|owi)$'
+
+    def __init__(self, db_manager: DBManager):
+        self.db_manager = db_manager
+
+        self.elastic_search_manager = ElasticsearchManager()
+
+        self.elastic_search_manager.createElasticConnection()
+        self.elastic_search_manager.createElasticSearchIngestPipeline()
+        self.elastic_search_manager.createElasticSearchIndex()
+
+        self.constants = get_constants()
+
+    def cluster_record(self, record_uuid):
+        query_filters = [
+            Record.frbr_status == 'complete',
+            Record.cluster_status == False,
+            Record.source.notin_(['oclcClassify', 'oclcCatalog']),
+            Record.title.isnot(None),
+            Record.uuid == record_uuid
+        ]
+
+        get_unclustered_record_query = self.db_manager.session.query(Record).filter(*query_filters)
+
+        unclustered_record = get_unclustered_record_query.first()
+        if not unclustered_record:
+            logger.info(f'No unclustered record found with UUID {record_uuid}')
+            return
+
+        try:
+            work, stale_work_ids = self._get_clustered_works(unclustered_record)
+            self._update_elastic_search(works_to_index=[work], works_to_delete=stale_work_ids)
+            self._delete_stale_works(stale_work_ids)
+            logger.info(f'Clustered record with uuid: {record_uuid}')
+        except Exception as e:
+            logger.exception(f'Failed to cluster record {unclustered_record}')
+            raise e
+
+    def _get_clustered_works(self, record: Record):
+        matched_record_ids = self._find_all_matching_records(record) + [record.id]
+
+        clustered_editions, records = self._cluster_matched_records(matched_record_ids)
+        work, stale_work_ids = self._create_work_from_editions(clustered_editions, records)
+
+        try:
+            self.db_manager.session.flush()
+        except Exception:
+            self.db_manager.session.rollback()
+            logger.exception(f'Unable to cluster record {record}')
+
+        self._update_cluster_status(matched_record_ids)
+
+        return work, stale_work_ids
+
+    def _update_elastic_search(self, works_to_index: list, works_to_delete: set):
+        self.elastic_search_manager.deleteWorkRecords(works_to_delete)
+        self._index_works_in_elastic_search(works_to_index)
+
+    def _delete_stale_works(self, work_ids: set[str]):
+        self.deleteRecordsByQuery(self.db_manager.session.query(Work).filter(Work.id.in_(list(work_ids))))
+
+    def _cluster_matched_records(self, record_ids: list[str]):
+        records = self.db_manager.session.query(Record).filter(Record.id.in_(record_ids)).all()
+
+        kmean_manager = KMeansManager(records)
+
+        kmean_manager.createDF()
+        kmean_manager.generateClusters()
+        editions = kmean_manager.parseEditions()
+
+        return editions, records
```

Comment:
I think the fulfill process will want the records back that have been clustered!